runtime/sema/type.go

```diff
@@ -4542,6 +4563,63 @@ func (t *InterfaceType) FieldPosition(name string, declaration *ast.InterfaceDec
 	return declaration.Members.FieldPosition(name, declaration.CompositeKind)
 }
 
+func (t *InterfaceType) EffectiveInterfaceConformances() []Conformance {
+	t.effectiveInterfaceConformancesOnce.Do(func() {
+		t.effectiveInterfaceConformances = distinctConformances(
+			t.ExplicitInterfaceConformances,
+			nil,
+			map[*InterfaceType]struct{}{},
+		)
+	})
+
+	return t.effectiveInterfaceConformances
+}
+
+// distinctConformances recursively visit conformances and their conformances,
+// and return all the distinct conformances as an array.
+func distinctConformances(
+	conformances []*InterfaceType,
+	parent *InterfaceType,
+	seenConformances map[*InterfaceType]struct{},
+) []Conformance {
+
+	collectedConformances := make([]Conformance, 0)
```

Comment:
Maybe avoid allocating empty slices and pre-allocate with capacity:

```suggestion
	if len(conformances) == 0 {
		return nil
	}

	collectedConformances := make([]Conformance, 0, len(conformances))
```

Reply:
Added in https://github.com/onflow/cadence/pull/2112/commits/c1486dbc47b13a5181c1a94e44767ad0a6a61afa. However, can't set the array capacity, since this is a recursive call and the size is unknown.